### ISSUE TITLE
x/sys/unix: Definition for O_PATH missing for most FreeBSD architectures

### ISSUE THREAD

**Issue by dfr** [golang/go#54355]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.18.2 freebsd/amd64
</pre>

### Does this issue reproduce with the latest release?

yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/dfr/.cache/go-build"
GOENV="/home/dfr/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="freebsd"
GOINSECURE=""
GOMODCACHE="/home/dfr/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="freebsd"
GOPATH="/home/dfr/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/freebsd_amd64"
GOVCS=""
GOVERSION="go1.18.2"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="cc"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/home/dfr/go/src/github.com/containers/podman/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3986639741=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

import "golang.org/x/sys/unix" and try to use unix.O_PATH on FreeBSD 13.1 or later

### What did you expect to see?

The program should compile 

### What did you see instead?

the program does not compile

### Notes

This symbol has been present in FreeBSD system headers since 13.1-RELEASE. The generated go code in x/sys/unix contains a definition for riscv in zerrors_freebsd_riscv64.go but the constant is missing for all other architectures. This symbol is not arch-dependent and should have the same value on all supported architectures.

**Comment by dmgk** [comment]
I think adding O_PATH definition was a bit premature, given that the minimum supported FreeBSD version in go1.20 is 12.3, which lacks O_PATH support.

**Comment by dmgk** [comment]
cc @golang/freebsd

**Comment by dfr** [comment]
As far as I can tell, attempting to use the O_PATH flag on 12.3 will just ignore it and open the directory as a file which probably isn't ideal. In my case, I only plan to use thie in FreeBSD 13.1 or later but perhaps pubishing the symbol in x/sys/unix isn't ideal if 12.x is a supported release. The value for O_PATH does not conflict with any other flags supported in 12.x.

My workaround is to define the O_PATH constant in the application code which is a bit of a mess but otherwise should work.

**Comment by dmgk** [comment]
My concern is that writing/testing application on 13.1 and then using it on 12.3 might create a false assumption that it implements something that in fact is not supported by the host OS.

**Comment by paulzhol** [comment]
I'm with agreement with @dmgk here, the zerrors_freebsd_riscv64.go should have been generated under a 12.3 Jail.

**Comment by dfr** [comment]
Makes sense - thanks for the added context. When my work-in-progress (podman on FreeBSD) makes it to the FreeBSD ports/package tree, I'll make sure its only built for 13.1 or later.

**Comment by ayang64** [comment]
@paulzhol  -- all of the 12.x releases will probably be EOL'd [December 2023](https://www.freebsd.org/security/#sup).

would it be reasonable to include it's definition at that time and close this issue?

**Comment by dfr** [comment]
That seems reasonable. I guess this would change all the references to stable/12 in mkall.sh to stable/13 which would pick up any other additions to the ABI post 12.x

**Comment by mlsgit** [comment]
As I am trying to get FreeBSD netlink into go (see go-nuts list about this): when is the move to FreeBSD 13 as reference branch planned? 

And is there anything to help?

**Comment by dfr** [comment]
I think it just needs someone to submit a pull request. FreeBSD-12 reached EOL at the end of 2023 and the current oldest supported version is 13.3 so the go reference branch could reasonably move to 13.3 (which should have working netlink)

**Comment by mlsgit** [comment]
I will see what I can do, i.e., I'm at to change it to 13.3.

Am 06.09.24 um 17:04 schrieb Doug Rabson:

**Comment by dotwaffle** [comment]
Just a heads up on this: I just tried to use `O_PATH` in FreeBSD, and found it wasn't there... Interestingly it *is* there for [riscv64](https://cs.opensource.google/go/x/sys/+/master:unix/zerrors_freebsd_riscv64.go;l=1237) though!

Is someone working on this already, or should I start work afresh on this (noting everything in https://cs.opensource.google/go/x/sys/+/master:unix/README.md)
